### DISCUSSION TITLE
Support `graphql` backticks syntax

### DISCUSSION THREAD

**Discussion by nandorojo** [dotansimha/graphql-code-generator#8584] [Ideas] [answered]
I'm trying to use this [VSCode plugin](https://marketplace.visualstudio.com/items?itemName=GraphQL.vscode-graphql) to get autocomplete with GraphQL.

However, it requires the following syntax:

```tsx
const query = graphql`
  query Test {
    test
  }
`
```

Notice that `graphql` has the backticks rather than parentheses. 

Could the `client-preset` add support for this? It would be easy to implement. It just needs to accept either a string or an array with the string.

```ts
// before
export function graphql(source: string | [string]) {
  return (documents as any)[source] ?? {}
}

// after
export function graphql(source: string | [string]) {
  if (Array.isArray(source)) {
    return (documents as any)[source[0]] ?? {}
  }
  return (documents as any)[source] ?? {}
}
```

And for the codegen'd function overloads, instead of doing this:

```ts
function graphql(source: '\n  query Test2($key: String!) {\n    sentryTesting(key: $key)\n  }\n'): ...
```
...it could do this:

```ts
// separated for legibility
type Source = '\n  query Test2($key: String!) {\n    sentryTesting(key: $key)\n  }\n'

function graphql(...source: [Source] | [[Source]]): ...
```

**Comment by saihaj** [comment, marked as answer]
Have you tried the GraphQL magic comment

```ts
const query = graphql(/* GraphQL */`
  query Test {
    test
  }
`)
```

**Reply by nandorojo** [reply]
Yeah I’m using that now. It works, but it’s not as nice of an API to need to include a comment.

**Reply by n1ru4l** [reply]
@nandorojo unfortunately there is a TypeScript limitation that makes it impossible to infer the correct type using tagged template literals.

**Reply by nandorojo** [reply]
I see, I didn't know that. In that case, it would be nice to add the current syntax to the vscode plugin

**Reply by revero-doug** [reply]
@nandorojo I can confirm the current syntax works with GraphQL: Language Feature Support (`graphql.vscode-graphql`) (but not Apollo GraphQL which I'd prefer to use for Apollo Studio integration). See https://www.the-guild.dev/graphql/config for how to make sure the plugin picks up your schema. if you maintain/generate your schema to live at `./schema.graphql` (addressed from project root), your config could be as simple as this (saved at ./graphql.config.yaml):

```yaml
schema: ./schema.graphql
```

**Reply by nandorojo** [reply]
@revero-doug it works without the graphql comment?

**Reply by revero-doug** [reply]
@nandorojo  no - the graphql comment should be thought of as part of the current syntax

**Reply by nandorojo** [reply]
As a quick update, it seems like this now works without the comment tag for me. Not sure if anything changed, but I get autocomplete now.

**Comment by revero-doug** [comment]
@nandorojo have you tried modifying the generated code as you suggest? tl;dr it won't return a TypedDocumentNode because typescript's type system can't (currently) support overloads based on TemplateStringArray contents (or if it can, someone with type-fu greater than mine will need to chime in).

I agree with your stance that the tagged template literal syntax is preferable, and I'd go a step further and say it's the idiomatic way inline gql is written.

Finding a workaround for this should be explicitly added to the roadmap IMO, and if it's deemed impossible, this caveat should be clearly documented in the official docs.

**Reply by nandorojo** [reply]
Ahh, sounds like I have my answer then. I didn’t realize overloads with template literal arguments aren’t supported.
